pubpub/pubpub
# Allow the SDK to use PubPub as a submodule

Terminology
- SDK: the thing at https://github.com/tefkah/pubpub-client
- "the client": the thing generated by `initClient` in `utils/api/client.ts` 
___

This PR allows PubPub v6 to be used as a submodule in the SDK, see this PR for in depth details on how that works: https://github.com/tefkah/pubpub-client/pull/7

Relevant for the current PR, there are a couple of things that needed to be changed on v6's side in order to accommodate this. Most of these changes where necessary in order to spin up a version of v6 to be used in the tests of the sdk, that way the sdk

## 🔥 Replaced the `webpackConfig.dev.js` loader from `ts-loader` to `esbuild-loader`, leading to `8x` faster compile times

In order to run the tests properly, the client needs to be compiled. Current compilation is really slow, taking 45 seconds on my machine. 

To speed this up, I swapped out the transpiler from `ts-loader` to `esbuild-loader`, which as the name might suggest uses `esbuild` instead of `tsc` to load the typescript/js files, which is significantly faster.

This, surprisingly, worked almost immediately! The only changes I had to make were to re-export typescript types, like here:

https://github.com/pubpub/pubpub/blob/c7c8c7033224896523c786bc042f639b657a371b/client/components/Editor/plugins/code/index.ts#L29-L39

`esbuild` does not do typechecking, so unless explicitly told to ignore these by adding `type` in front, it will assume these are javascript functions/objects/etc., and will complain (warn) that it cannot find these.
Simply adding the `type` annotation in front solves this.


### Downsides

There is one downside about this approach, namely that `esbuild` does not do any typechecking.

I do not think this is a problem, because the whole codebase should be typechecked anyway before publishing. That's why I instead included `tsc --noEmit` at the end of `npm run lint`, which always gets run in CI. This way we know for sure that the types are alright, and still save a lot of time locally developing.

Also, I did not change this in the prod config anyway, just to be safe!


## Moving some stuff around/changing imports

In order to have imports resolve correctly when running the app from a different directory and to avoid importing things incredibly deeply in the SDK, I moved some things around a bit. 

These fall into X categories:

1. Moving inline types from e.g. `client` to `types`
2. Copied some enums (which were actual JS, not types) from `types` to the schema where I need them. This allows `tsup` to kind of ignore `types` for bundling the code, which lessens the possibility of accidentally including a lot v6 code in the SDK.
3. Changed some explicit `dist` imports. The were like `require(process.cwd() + '/dist')`. If you run the app from the sdk (which is one dir up), this would obviously not work, so I changed it to something like `require(__dirname + '/dist')`, so it's relative to the file instead. This does not always work, because sometimes the files are themselves (assumed to be) in `dist`, so if that was the case I did something not very pretty, like

https://github.com/pubpub/pubpub/blob/c7c8c7033224896523c786bc042f639b657a371b/server/Html.tsx#L7-L11

## Changed all `sequelize-typescript` class definitions to use `declare` for the fields instead of just putting them in there.
See e.g. here: https://github.com/pubpub/pubpub/blob/c7c8c7033224896523c786bc042f639b657a371b/server/collection/model.ts#L48-L65


To be quite honest I'm not entirely sure why this was needed. When running tests in v6, everything is fine, but when running them through the submodule in the sdk, I get errors like these:
```log
  console.warn
      (sequelize) Warning: Model "Pub" is declaring public class fields for attribute(s): "id", "slug", "title".
      These class fields are shadowing Sequelize's attribute getters & setters.
      See https://sequelize.org/main/manual/model-basics.html#caveat-with-public-class-fields
```

This indicates that sequelize is overwriting already declared fields somehow. This not only leads to these warnings, but to the whole app not working.

Changing all fields to use `declare` instead does work, and does not seem to negatively impact the core app, so that's why I just ran with it. This might be because I am running the tests in the `sdk` with `swc` to speed them up, idk.

I do not think this will pose a problem in the future, e.g. when someone would forget to add `declare` somewhere. As stated, the core app works fine without it, it might break a test in the sdk maybe, but the `sdk` does not need the entire app to be able to run to work, only to test it.
  


## Issue(s) Resolved

#2828 
#2851 

## Test Plan

### To test SDK

1. Pull this PR: https://github.com/tefkah/pubpub-client/pull/7 with `--recurse-submodules`
2. `pnpm i`
3. `pnpm run ci` to see the tests work.

### To test that the app has no regressions

I think a smoke test is in order. I'm quite confident nothing of note has changed, but it's good to check.
